In [2]:
# %matplotlib inline
# import matplotlib.pyplot as plt
# import seaborn as sns
import numpy as np
import pandas as pd
# from IPython.display import display
# from IPython.display import Image
# sns.set_context("notebook")

In [3]:
# from google.colab import drive
# drive.mount('/content/drive')

# Extracting Data and Filtering
:The `[]` operator supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows.

The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

# Q1.
Select multiple criteria by creating multiple boolean Series and combining them using the `&` operator. For Example Result is win and %win is more than 50.

In [4]:
orgelections = pd.read_csv("datafiles/elections1.csv") # reading a CSV file into a data frame
# if we end a cell with an expression or variable name, the result will print the data frame
orgelections

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

We will now work with a dataset containing the names of all babies born in California, USA. Visit the [Social Security Adminisitration Website](https://www.ssa.gov/oact/babynames/limits.html) for more details about the dataset if you want.

**Let's start by loading the Baby Names dataset.**

In [5]:
# import urllib.request
# import os.path
# import zipfile
# import pandas as pd

# # data_url = "https://www.ssa.gov/oact/babynames/state/namesbystate.zip"
# # local_filename = "babynamesbystate.zip"
# # if not os.path.exists(local_filename): # if the data exists don't download again
# #     with urllib.request.urlopen(data_url) as resp, open(local_filename, 'wb') as f:
# #         f.write(resp.read())

# zf = zipfile.ZipFile("files/babynamesbystate.zip", 'r')

# ca_name = 'STATE.CA.TXT'
# field_names = ['State', 'Sex', 'Year', 'Name', 'Count']
# with zf.open(ca_name) as fh:
#     babynames = pd.read_csv(fh, header=None, names=field_names)

# babynames.head(15)

In [1]:
import zipfile

# Specify the path to your ZIP file using forward slashes
zip_file_path = "datafiles/babynamesbystate.zip"

# Specify the name of the file you want to read from the ZIP archive
ca_name = 'STATE.CA.TXT'

# Open the ZIP file
with zipfile.ZipFile(zip_file_path, 'r') as zf:
    # Check if the specified file exists in the ZIP archive
    if ca_name in zf.namelist():
        # Read the CSV file directly from the ZIP archive
        with zf.open(ca_name) as fh:
            # Define field names for the DataFrame
            field_names = ['State', 'Sex', 'Year', 'Name', 'Count']
            # Read CSV using Pandas
            orgbabynames = pd.read_csv(fh, header=None, names=field_names)
            # Display the first 15 rows of the DataFrame
            # orgbabynames.head(15)
    else:
        print(f"File '{ca_name}' not found in the ZIP archive.")

# orgbabynames.head(15)
babynames = orgbabynames.copy()
babynames.head(10)

NameError: name 'pd' is not defined

# Q2.
Calculate the some statics of babynames DataFrame using Utility Functions, like .describe(), .size(), .shape() and report your analysis in comments

In [7]:
# first checking the info
babynames.info()
# checking the numerical discription
babynames.describe()
# checking shape and size
# babynames.shape
# babynames.size

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 407428 entries, 0 to 407427
Data columns (total 5 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   State   407428 non-null  object
 1   Sex     407428 non-null  object
 2   Year    407428 non-null  int64 
 3   Name    407428 non-null  object
 4   Count   407428 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 15.5+ MB


Year          Count
count  407428.000000  407428.000000
mean     1985.733609      79.543456
std        27.007660     293.698654
min      1910.000000       5.000000
25%      1969.000000       7.000000
50%      1992.000000      13.000000
75%      2008.000000      38.000000
max      2022.000000    8260.000000

### Info():
- total enteries
- there are 5 columns
- default row indexing
- dtypes in dataframe
- total observartions in serires and no null values
### describe():
- describing the numerical data: `descriptive stats of numeric series`
### Size:
- There are total `2037140` entries in dataframe
### Shape:
- There are `407428` observations/rows and 5 `attributes/columns/series` in dataframe 

# Q.3:

Using the babynames DataFrame that contains baby names data for the state of California, how can you use the 'isin' method to find out which records have names that are in a given list of names, such as "Emma," "Liam," and "Olivia," and what would be the output of this operation when applied to the first 15 rows of the DataFrame (babynames.head(15))?

In [8]:
# List of names to check
name_list = ["Emma", "Liam", "Olivia"]

# Use the isin method to filter records with names in the name_list
# on complete dataframe:
# names3 = babynames[babynames['Name'].isin(name_list)] # conditions always in small brac.
# # names3
obs15 = babynames.head(15)
names3 = obs15[obs15['Name'].isin(name_list)] # conditions always in small brac.

# Display the filtered DataFrame
# obs15
names3



Empty DataFrame
Columns: [State, Sex, Year, Name, Count]
Index: []

- no names in first 15 rows were found!

# Q.4:
How can you use the 'str.startswith' method in pandas to filter records in a babyname DataFrame? Provide an example code using the babynames DataFrame and explain how this method can be employed to extract data for baby names that begin with 'A' or 'B' set of letters.

In [9]:
# Filter rows where the 'Name' column starts with 'M'
names_with_m = babynames[(babynames['Name'].str.startswith('M'))]
# Display the filtered DataFrame
names_with_m

names_ab = babynames[(babynames['Name'].str.startswith('A')) | (babynames['Name'].str.startswith('B'))]
# print("Baby names starting with 'A' and B:")
names_ab


State Sex  Year     Name  Count
7         CA   F  1910    Alice    118
23        CA   F  1910  Bernice     59
31        CA   F  1910     Anna     51
32        CA   F  1910      Ann     47
36        CA   F  1910    Betty     40
...      ...  ..   ...      ...    ...
407125    CA   M  2022   Brando      5
407126    CA   M  2022  Braxtyn      5
407127    CA   M  2022  Brenden      5
407128    CA   M  2022     Brix      5
407129    CA   M  2022    Broly      5

[61691 rows x 5 columns]

# Q.5

Find the most popular baby name in California in 2012

__Exercise__: Try to find the most popular baby name by finding the maximum value for 'Count' and then apply filtering to find the name.


In [10]:
# babynames
popname_2012 = babynames[(babynames['Year'] == 2012)]
popname_2012 = popname_2012.sort_values(by= 'Count', ascending=False)
popname_2012
popname_2012.loc[197530,['Count','Name']]


Count      3644
Name     Sophia
Name: 197530, dtype: object

# Q. 6
Find baby names that start with J, in th Year from 2001 to 2019

*Note: To answer this question, we need a way to look at the individual characters of the baby names*

In [11]:
# Note: Series has powerful string manipulation methods associated with it
names_with_j = babynames[(babynames['Name'].str.startswith('J')) & 
                         (babynames['Year'] >= 2001) &
                         (babynames['Year'] <= 2019)]
names_with_j

State Sex  Year         Name  Count
152826    CA   F  2001      Jessica   2244
152828    CA   F  2001     Jennifer   2027
152832    CA   F  2001      Jasmine   1655
152847    CA   F  2001   Jacqueline   1195
152857    CA   F  2001        Julia    983
...      ...  ..   ...          ...    ...
398685    CA   M  2019  Joseantonio      5
398686    CA   M  2019       Jovian      5
398687    CA   M  2019       Jujhar      5
398688    CA   M  2019      Jullian      5
398689    CA   M  2019         June      5

[11276 rows x 5 columns]

# Q.7
Add a new column called "name_length" to the babynames DataFrame, you can use .str.len() in existing Name column. Moreover. also modify the Length values with new values like existing_value-1 and rename the length column with new label as desceribe in slides.



In [12]:
# Calculate the length of each name and add it as a new column "_length"

# calculating the length of each name
name_length = babynames['Name'].str.len()
# name_length

# now adding series to the dataframe
babynames['Name_length'] = name_length
# babynames

#modifying the name_length column by adding 1
babynames['Name_length'] += 1
# again -1
babynames['Name_length'] -= 1

# Display the updated DataFrame
babynames

# Rename the "_length" column to "Length"
babynames = babynames.rename(columns= {'Name_length':'Length'})
babynames

State Sex  Year      Name  Count  Length
0         CA   F  1910      Mary    295       4
1         CA   F  1910     Helen    239       5
2         CA   F  1910   Dorothy    220       7
3         CA   F  1910  Margaret    163       8
4         CA   F  1910   Frances    134       7
...      ...  ..   ...       ...    ...     ...
407423    CA   M  2022   Zayvier      5       7
407424    CA   M  2022       Zia      5       3
407425    CA   M  2022      Zora      5       4
407426    CA   M  2022    Zuriel      5       6
407427    CA   M  2022      Zylo      5       4

[407428 rows x 6 columns]

# Q.8

Use .str.contains() to sample the names with 'ad' string and show first 10 samples.

In [13]:
names_have_ad = babynames[(babynames['Name'].str.contains('ad'))]
sample_ad10 = names_have_ad.sample(10, replace=True)
sample_ad10

State Sex  Year       Name  Count  Length
380026    CA   M  2013       Brad     11       4
329875    CA   M  1994       Emad      5       4
278480    CA   M  1966       Brad    153       4
276579    CA   M  1964      Grady     13       5
160566    CA   F  2003  Guadalupe    517       9
406328    CA   M  2022  Guadalupe      9       9
224515    CA   F  2018      Brady      5       5
263491    CA   M  1952       Wade     34       4
179097    CA   F  2007      Kadyn     10       5
400515    CA   M  2020       Hadi     10       4

More on str methods when we'll study regular expressions in this course. Note that there are other approaches that we can use to find the same answer as above.

In [14]:
# Create a boolean mask for rows containing 'ad' in the "Name" column
bool_names_have_ad = (babynames['Name'].str.contains('ad'))
mask = bool_names_have_ad

# Use the mask to select rows and sample 5 random rows
population_ad10 = babynames['Name'][mask]
population_ad10 = population_ad10.to_frame()
sample2_ad10 = population_ad10.sample(10, replace=True)
sample2_ad10


Name
729        Sadie
378806  Salvador
254487      Wade
302950      Chad
392331   Gennady
195814   Jadelyn
207249    Maddie
299834     Ahmad
349592    Maddox
338604      Jade

# Q.9
 Sort names by their length.

In [15]:
# sorting by name length:
babynames_bylen = babynames.sort_values(by='Length')
babynames_bylen

State Sex  Year             Name  Count  Length
326570    CA   M  1993               An      8       2
292150    CA   M  1976               Al     13       2
252556    CA   M  1937               Al     21       2
401470    CA   M  2020               Jr      5       2
260022    CA   M  1948               Ed     43       2
...      ...  ..   ...              ...    ...     ...
339472    CA   M  1998  Franciscojavier      6      15
327358    CA   M  1993  Johnchristopher      5      15
337477    CA   M  1997  Ryanchristopher      5      15
312543    CA   M  1987  Franciscojavier      5      15
321792    CA   M  1991  Ryanchristopher      7      15

[407428 rows x 6 columns]

In the babynames DataFrame, which baby name appears most frequently across all years and states, and how can you use pandas to find this information? Provide the code and steps to determine the most popular baby name in the dataset.

In [16]:
# Group the data by "Name" and sum the "Count" for each name
grp_name_count = babynames[["Name", "Count"]].groupby("Name").agg(sum)

# Find the name with the highest total count
grp_name_count = grp_name_count.sort_values(by='Count', ascending=False)
grp_name_count

# Display the most popular name and its count
grp_name_count.iloc[0]

C:\Users\Maaz PC\AppData\Local\Temp\ipykernel_16720\3457991876.py:2: FutureWarning: The provided callable <built-in function sum> is currently using DataFrameGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  grp_name_count = babynames[["Name", "Count"]].groupby("Name").agg(sum)


Count    435716
Name: Michael, dtype: int64

# Q.10

1: How can you use the .sample() method to randomly select a single row from the babynames DataFrame, and what is the purpose of such a selection?

2: Suppose you want to generate a random sample of 10 baby names from the babynames DataFrame. How would you use the .sample() method to achieve this, and what parameters or arguments would you specify?

3: In a scenario where you need to ensure reproducibility, explain how you can set a random seed when using the .sample() method in pandas. Why might this be important in certain data analysis tasks?

4: You want to perform a bootstrap resampling analysis on the baby names data in the babynames DataFrame. Describe how you can repeatedly sample with replacement using .sample() to create multiple bootstrap samples for statistical analysis

In [19]:
# # part 1 and 2:
random_sample = babynames.sample(n=10, replace=True)
random_sample

# part 3:
random_sample = babynames.sample(n=10, replace=True, random_state= 5)
random_sample

# 4th Part
# Assuming you have the 'babynames' DataFrame
babynames_count = babynames['Count']
# Set the random seed for reproducibility
# np.random.seed(5)
# Number of bootstrap samples
bootstrap_samples = 1000

# Sample size (same as the original dataset)
sample_size = len(babynames_count)
# Initialize a list to store bootstrap sample statistics
bootstrap_mean = []

# Perform bootstrapping
for i in range(bootstrap_samples):
    # Generate a bootstrap sample with replacement
    bootstrap_sample = babynames_count.sample(n=sample_size, replace=True)
    # Perform your statistical analysis on the bootstrap sample here
    mean = babynames_count.mean()
    # For example, you can calculate the mean, median, or other statistics
    # Append the statistic of interest to the list
    bootstrap_mean.append(mean)

bootstrap_mean

[79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,
 79.54345553079317,


## Ans
### 1
- The purpose is to obtain a random sample of data
- Every time a new sample becomes the representative of dataframe
- This can be useful for various purposes such as exploratory data analysis
- creating a representative subset of data for modeling
- randomizing the order of data for analysis.
- fiding the deviatin of data from mean to observe different trends
### 2
- Setting a random seed ensures that the random sampling process generates the same results each time
- It is important for reproducibility in data analysis tasks. 
- This is particularly useful when sharing code or results with others or when needing to reproduce an analysis.
### 3
- we can reproducibility by adding a parameter of 'random_state', it ensures that the random sampling is reproducible.
- it is useful in sharing code, when someone else want to run and ensure the the validation and work on the previous building 


# Q.11
1: The code randomly mixes up the list of baby names. Why might someone want to do this, and how could it affect the analysis of the names?

2: After mixing, the code figures out how long each baby name is. Why would we care about name lengths, and what can this tell us about the names?

3: The code arranges the name lengths from longest to shortest. Why might we want to do this, and what could it help us find in the data?

4: The code finds the order of baby names based on their lengths and shows us the first few names in this new order. When might we need to organize names this way, and what might we discover from the first few names in this order?

In [37]:
#Part1, shuffling can be done using sample(frac=1)
shuffled_names = babynames.sample(frac=1)
# shuffled_names

#Part2
# Calculate the length of each baby name
babynames['name_lengths'] = babynames['Name'].str.len()
# babynames

#Part3
# Sort the name lengths from longest to shortest
sorted_names = babynames.sort_values(by= 'name_lengths', ascending=False)
# sorted_names

#Part4
# Find the order of baby names based on their lengths
# Show the first few names in this new order
sorted_names.head(15)



State Sex  Year             Name  Count  Length  name_lengths
334166    CA   M  1996  Franciscojavier      8      15            15
337301    CA   M  1997  Franciscojavier      5      15            15
339472    CA   M  1998  Franciscojavier      6      15            15
321792    CA   M  1991  Ryanchristopher      7      15            15
327358    CA   M  1993  Johnchristopher      5      15            15
321913    CA   M  1991  Franciscojavier      6      15            15
344440    CA   M  2000  Franciscojavier      6      15            15
327472    CA   M  1993  Ryanchristopher      5      15            15
313977    CA   M  1988  Franciscojavier     10      15            15
316886    CA   M  1989  Franciscojavier      6      15            15
312543    CA   M  1987  Franciscojavier      5      15            15
337477    CA   M  1997  Ryanchristopher      5      15            15
102505    CA   F  1986  Mariadelosangel      5      15            15
147522    CA   F  1999   Mariaguadalupe     10      14            14
209419    CA   F  2014   Mariaguadalupe      5      14            14

### 1
- It can remove any inherent bias or ordering effects in the original list.
- Mixing the list ensures that each name has an equal chance of being selected or analyzed. 
### 2
- Knowing the lengths of baby names can provide insights into naming preferences and trends. 
- Longer names might indicate cultural or regional influence.
- shorter names might be more traditional or popular in certain demographics. 
- Analyzing name lengths can help researchers understand the diversity and complexity of naming practices within a population.
### 3
- Arranging the name lengths from longest to shortest can help identify outliers and patterns in the data. 
- It allows researchers to see the distribution of name lengths more clearly and identify any unusual or unexpected lengths. 
- This sorting can also reveal trends such as a preference for longer or shorter names within the dataset.
### 4
- Organizing names based on their lengths can be useful for various purposes, such as exploring naming trends, identifying popular name lengths, or conducting statistical analysis. 
- Examining the first few names in this new order can provide insights into common name lengths and potentially reveal patterns or trends in naming conventions. For example, if the first few names are all relatively short or long, it may indicate a preference for names of a certain length within the dataset.